scipy/spatial/ckdtree/src/build.cxx

```diff
@@ -153,51 +195,97 @@ build(ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
             tmp_mins[d] = fixed_val;
             tmp_maxes[d] = fixed_val;
 
-            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact);
+            return build(self, start_idx, end_idx, tmp_maxes, tmp_mins, _median, _compact, buflist, workers);
         }
 
+        build_worker worker;
+
         if (CKDTREE_LIKELY(_compact)) {
-            _less = build(self, start_idx, p, maxes, mins, _median, _compact);
-            _greater = build(self, p, end_idx, maxes, mins, _median, _compact);
+            workers -= worker.run(self, p, end_idx, maxes, mins, _median, _compact, workers);
+            _less = build(self, start_idx, p, maxes, mins, _median, _compact, buflist, workers);
+            if(!worker)
+                _greater = build(self, p, end_idx, maxes, mins, _median, _compact, buflist, workers);
```

Comment:
I did think about `reserve` and tried it: The effect is not noticable. `std::vector` scales up the allocation sizes each time it has to resize, such that for large trees it does even become a bit faster than `deque`.

In general, I think `reserve` would be a good idea, because on some systems it may provide an advantage. I was reluctant to add `reserve` because it was not there before and had the same thought as @sturlamolden  about the memory overhead: in an imbalanced tree, using `reserve` would allocate more memory than the sequential implementation.

Reply:
> How about instead of modifying this to handle both single and multithreaded, you could move the partitioning code into a helper function and have two versions of build.

If the PR would be accepted this way, I could do that. I was going for a minimal patch size in my implementation. The version with one `build` function would also be more compatible with an approach with dynamic load balancing, because the the fanning-out would not be static.

Reply:
The reserve is in https://github.com/scipy/scipy/commit/fbb4a994ae81441c09baa299d8f9b0c1ce54f4ef . I am still concerned, that the two build functions would duplicate too much code.

Reply:
>I am still concerned, that the two build functions would duplicate too much code.

This is what I expect `build_parallel` would look like:
```cpp
static ckdtree_intp_t
build_parallel(
      ckdtree *self, ckdtree_intp_t start_idx, intptr_t end_idx,
      double *maxes, double *mins, const int _median, const int _compact
      std::list<std::vector> *buflist, int workers)
{
    auto &node_buffer = buflist->back();
    const int min_parallel_leafs = /* Some minimum grain size */
    if (end_idx-start_idx <= self->leafsize * min_parallel_leafs) {
        return build(..., node_buffer);
    }

    auto res = split_and_partition(start_idx, end_idx, maxes, mins, _median, _compact);
    ckdtree_intp_t d = res.d;
    ckdtree_intp_t p = res.p;
    double split = res.split;

    auto node_index = allocate_node(node_buffer, start_idx, end_idx);
    
    if (CKDTREE_UNLIKELY(d == -1)) {
        // return leaf node
        ckdtree* n = &node_buffer[node_index];
        n->split_dim = -1;
        return node_index;
    }

    if (CKDTREE_LIKELY(_compact)) {
        /* launch one of these in parallel */
        build_worker worker(...);
        _greater = build_parallel(...);
        _less = worker.future.get();
    }
    else
    {
        std::vector<double> mids(maxes, maxes + m);
        mids[d] = split;
        build_worker worker(...);

        for (i=0; i<m; ++i) mids[i] = mins[i];
        mids[d] = split;
        _greater = build(self, p, end_idx, maxes, mids, _median, _compact);
        _less = worker.future.get();
    }

    
    /* fill in entries */
    ckdtreenode *n = &node_buffer[node_index];
    n->_less = _less;
    n->_greater = _greater;
    n->split_dim = d;
    n->split = split;

    return node_index;
}
```

Yes there is necessarily duplicates, but most of the heavy lifting has been factored into `split_and_partition` which determines the split point and partitions the indices.

Reply:
This case cannot be factored-out:
https://github.com/scipy/scipy/blob/cd4029534163a43d2b6679578c5f60322e88a002/scipy/spatial/ckdtree/src/build.cxx#L203-L204
as-is, because it calls `build`. The correct `build` would need to passed as a pointer or a template predicate, at lease the first one would require `split_and_partition` the `buflist` and `workers` arguments.

Is there anything speaking against having a loop instead of the recursive call?
```c++
do {
 // partitioning
 //....

 if (CKDTREE_UNLIKELY(p == start_idx || p == end_idx)) {
   assert(!_compact);
   const auto fixed_val = data[indices[start_idx]*m + d];
   mins[d] = fixed_val; // tmp_mins is not needed as mins and maxs are modified in the  _compact case anyway,
   maxes[d] = fixed_val; // i.e. they are temporaies
 }
} while(CKDTREE_UNLIKELY(p == start_idx || p == end_idx));
```

Reply:
I pushed a split version using the loop. I took the liberty of also cleaning up the local variable declarations (declare on use, not Fortran-Style; and more const).